runtime/drivers/clickhouse/embed.go

```diff
@@ -0,0 +1,377 @@
+package clickhouse
+
+import (
+	"archive/tar"
+	"bufio"
+	"compress/gzip"
+	"fmt"
+	"io"
+	"net"
+	"net/http"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"runtime"
+	"strings"
+	"time"
+
+	"github.com/ClickHouse/clickhouse-go/v2"
+	"github.com/rilldata/rill/cli/pkg/dotrill"
+	"go.uber.org/zap"
+)
+
+const embedVersion = "24.7.4.51"
+
+type embedClickHouse struct {
+	tcpPort int
+	dataDir string
+	tempDir string
+	logger  *zap.Logger
+	cmd     *exec.Cmd
+}
+
+func newEmbedClickHouse(tcpPort int, dataDir, tempDir string, logger *zap.Logger) *embedClickHouse {
+	return &embedClickHouse{tcpPort: tcpPort, dataDir: dataDir, tempDir: tempDir, logger: logger}
+}
+
+// start installs (depending on OS and platform) and starts ClickHouse server.
+// The destination directory for the ClickHouse binary is .rill/clickhouse.
+// The function returns the DSN for the ClickHouse server and close function.
+func (e *embedClickHouse) start() (*clickhouse.Options, error) {
```

Comment:
Adding a todo here since this is important to handle in a follow up (will assign someone after this merges):

```suggestion
// The function returns the DSN for the ClickHouse server and close function.
// 
// TODO: Since this can be a long-running process, we should accept a `ctx`,
// but the `drivers.Open` function currently doesn't propagate that.
func (e *embedClickHouse) start() (*clickhouse.Options, error) {
```